### PR TITLE
easy way of initializing c++ tensor

### PR DESCRIPTION
**Pull request by ShubhamAgarwal12**
creating tesnor using numpy like initialization.
supports upto 3 dimension.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)

### CODE REVIEW

**Comment by yxlao** [comment]
As discussed, let's switch to the template-based impl: `Tensor::Init<T>({}, device);`

**Review by ssheorey — COMMENTED** [review]
Some tests with `Init<Bool>` (for advanced indexing) and `Init<UInt8>` (for images) would be useful too.

**Review by yxlao — CHANGES_REQUESTED** [review]
lgtm after resolving the comments

> Reviewed 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796#-:-MPMdRSZ8uMKUlAvOdf8:b-30qvkr)* status: all files reviewed, 10 unresolved discussions (waiting on @ShubhamAgarwal12)

___
*[cpp/open3d/core/Tensor.h, line 229 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPMZV2OEkxjI9ChHGYO:-MPMZV2P4_qPfax7_Zao:b-oz8m0u) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L229)):*
> ```objc
>                        const Device& device = Device("CPU:0")) {
>         Dtype type_ = Dtype::FromType<T>();
>         SizeVector shape_;
> ```

the `_` is reserved for class member variables, let's rename them.

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796#-:-MPecDXc4sEi4U9Xw9y5:bnfp4nl)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @ssheorey and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 222 at r2](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r2-222:-MPeEhIF0iHR7I6OLBex:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/1ca952cb95bfa37453a18e4cac509d62cfdd84c0/cpp/open3d/core/Tensor.h#L222)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

// comment -> /// comment for doxygen to pick up documentation
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 229 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPMZV2OEkxjI9ChHGYO:-MPeEhyNBLAVhYgIfxh_:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L229)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

the `_` is reserved for class member variables, let's rename them.
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 230 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-230:-MPeU7ADBdP2f7asvosV:b362ug3) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L230)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

Minor: Avoid `push_back` with `SizeVector shape_{static_cast<int64_t>(in_list.size())};`
</blockquote></details>


Done.
___
*[cpp/open3d/core/Tensor.h, line 240 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-240:-MPeEjFrC2tJ3iUdJNGl:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L240)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

/// doxygen comment
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 255 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-255:-MPeYjRg3nl42aq50Eav:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L255)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

"Inputs **rows** are of unequal size"
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 268 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-268:-MPeElFEEln_ODqEz82U:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L268)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

/// doxygen comment
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 286 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-286:-MPeZndm4lIZa0Z74mDC:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L286)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

"Inputs **columns** are of unequal size"
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 296 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-296:-MPe_PWdEdld6eHJ8BJE:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L296)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

"Inputs **rows** are of unequal size"
</blockquote></details>

Done.

**Comment by ShubhamAgarwal12** [comment]
> lgtm after resolving the comments
> 
> > Reviewed 2 of 3 files at r4.
> > _[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796#-:-MPMdRSZ8uMKUlAvOdf8:b-30qvkr)_ status: all files reviewed, 10 unresolved discussions (waiting on @ShubhamAgarwal12)
> 
> _[cpp/open3d/core/Tensor.h, line 229 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPMZV2OEkxjI9ChHGYO:-MPMZV2P4_qPfax7_Zao:b-oz8m0u) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L229)):_
> 
> > ```objc
> >                        const Device& device = Device("CPU:0")) {
> >         Dtype type_ = Dtype::FromType<T>();
> >         SizeVector shape_;
> > ```
> 
> the `_` is reserved for class member variables, let's rename them.

Done.

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: all files reviewed, 7 unresolved discussions (waiting on @ShubhamAgarwal12 and @ssheorey)

___
*[cpp/open3d/core/Tensor.h, line 255 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-255:-MPh1n053Na2BOLlU_KJ:b557wc7) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L255)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

In numpy, this is called "ragged" sequences (https://stackoverflow.com/q/63097829/1255535). It might be hard to tell `row` and `col` for higher dimensional tensors, so let's consider switching to a more generic description.

Something like:
```cpp
utility::LogError("Cannot create Tensor with ragged nested sequences (lists-of-lists with different lengths or shapes.)");
```
___
*[cpp/open3d/core/Tensor.h, line 286 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-286:-MPh2Q1i5lcXfy5p0RVZ:bhce3xh) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L286)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same, switch to the "ragged nested sequences" name.
___
*[cpp/open3d/core/Tensor.h, line 296 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-296:-MPh2WjlEqbMdRt0fNxL:bhce3xh) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L296)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same, switch to the "ragged nested sequences" name.
___
*[cpp/tests/core/Tensor.cpp, line 135 at r6](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPh2aq11nI-31cTPNqh:-MPh2aq11nI-31cTPNqi:bwkp67b) ([raw file](https://github.com/intel-isl/open3d/blob/25251b1aead2e079297e1fc86a53ca9bd3b9f924/cpp/tests/core/Tensor.cpp#L135)):*
> ```cpp
> 
>     // Check tensor element size mismatch.
>     EXPECT_THROW(core::Tensor::Init<int>({{1, 2, 3}, {4, 5}}, device),
> ```

Add more unit tests of special shapes with 0-sized dimensions:

```cpp
Tensor::Init<T>({});
Tensor::Init<T>({{}, {}});
Tensor::Init<T>({{}, {1}}); // exception
Tensor::Init<T>({{}, {{}}}); // exception
```

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @ssheorey and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 255 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-255:-MPi7dZ73gr5sLAXgrwk:b41km5n) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L255)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

In numpy, this is called "ragged" sequences (https://stackoverflow.com/q/63097829/1255535). It might be hard to tell `row` and `col` for higher dimensional tensors, so let's consider switching to a more generic description.

Something like:
```cpp
utility::LogError("Cannot create Tensor with ragged nested sequences (lists-of-lists with different lengths or shapes.)");
```
</blockquote></details>

True. I kept "elements" initially but thought it might be fine for up to 3D. I've changed to "Cannot create Tensor with ragged nested sequences( nested lists with unequal sizes or shapes)."
___
*[cpp/open3d/core/Tensor.h, line 286 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-286:-MPi8DyWCBg0ycqnHArt:bu15pxj) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L286)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Same, switch to the "ragged nested sequences" name.
</blockquote></details>

Same as earlier.
___
*[cpp/open3d/core/Tensor.h, line 296 at r4](https://reviewable.io/reviews/intel-isl/open3d/2796#-MP2pT362o-PHD0-Aqci-r4-296:-MPi8IKu-Y8LSFxIkiwJ:b-somrrr) ([raw file](https://github.com/intel-isl/open3d/blob/d6f2cea7287768f692e755120ad4a926982d23ba/cpp/open3d/core/Tensor.h#L296)):*
> Same as earlier.
___
*[cpp/tests/core/Tensor.cpp, line 135 at r6](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPh2aq11nI-31cTPNqh:-MPi8KeK1gH1K64X6F96:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/25251b1aead2e079297e1fc86a53ca9bd3b9f924/cpp/tests/core/Tensor.cpp#L135)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Add more unit tests of special shapes with 0-sized dimensions:

```cpp
Tensor::Init<T>({});
Tensor::Init<T>({{}, {}});
Tensor::Init<T>({{}, {1}}); // exception
Tensor::Init<T>({{}, {{}}}); // exception
```
</blockquote></details>

Done.

File: cpp/open3d/core/Tensor.h [hunk 1]
```diff
@@ -219,6 +219,104 @@ class Tensor {
...
+                            "Cannot create Tensor with ragged nested "
+                            "sequences( nested"
+                            "lists with unequal sizes or shapes).");
+                }
+            }
+            ele_list.insert(ele_list.end(), ele0.begin(), ele0.end());
+        }
+
+        SizeVector shape{dim0_size, dim1_size};
+        return Tensor(ele_list, shape, type, device);
```
Comment (ssheorey) [inline cpp/open3d/core/Tensor.h:262]: Can you check for zero size here too (similar to line 308)? eg `Init({{}})`: size = (1,0) convert to (1,)

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: 1 of 2 files reviewed, 12 unresolved discussions (waiting on @ShubhamAgarwal12, @ssheorey, and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 222 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJKF7BSdedMR0piGQ:-MPkJKF7BSdedMR0piGR:b-sae9ch) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L222)):*
> ```objc
>                        const Device& device = Device("CPU:0"));
> 
>     /// Create a 1-D tensor with given list.
> ```

Could you also add a new factory function for "Create a 0-D tensor"? 

In numpy, this is:
```python
np.array(3.14)
```
___
*[cpp/open3d/core/Tensor.h, line 254 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJg4t5XArzoMrwlep:-MPkJg4t5XArzoMrwleq:b-7pb8wb) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L254)):*
> ```objc
>                     utility::LogError(
>                             "Cannot create Tensor with ragged nested "
>                             "sequences( nested"
> ```

please reformat it to:`sequences (nested lists ...` (change space and line break location)
___
*[cpp/tests/core/Tensor.cpp, line 144 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJya6AwIuL-bj26rB:-MPkJya6AwIuL-bj26rC:b-qwr3xb) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/tests/core/Tensor.cpp#L144)):*
> ```cpp
>     // Testing shapes with 0-element.
>     std::vector<double> vals6({});
>     core::Tensor t6 = core::Tensor::Init<double>({}, device);
> ```

Assert the shape is `{0,}`.
___
*[cpp/tests/core/Tensor.cpp, line 148 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkKCwh6mVMz5lbltNQ:-MPkKCwh6mVMz5lbltNR:bawbd6n) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/tests/core/Tensor.cpp#L148)):*
> ```cpp
> 
>     std::vector<bool> vals7({});
>     core::Tensor t7 = core::Tensor::Init<bool>({{}, {}});
> ```

Assert the shape is `{2, 0}`.

Also, add shape and dtype assertion for other tests. The dtype assertion is needed to test if the template is converted correctly to core::Dtype.

File: cpp/open3d/core/Tensor.h (hunk 1 above)
Comment (yxlao) [reply cpp/open3d/core/Tensor.h:262]: Yes. Let's add a unit test for that.

In numpy, this is:
```python
In [5]: np.array([[]]).shape
Out[5]: (1, 0)
```

Comment (ShubhamAgarwal12) [reply cpp/open3d/core/Tensor.h:262]: I checked for this case and all the 0-sized initialiser list inputs are being passed to the function for 3-D. I added a test with "core::Tensor::Init<bool>({{}, {}})" for the 2-D case. I also added test with "core::Tensor::Init<bool>({{}})"

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: 0 of 2 files reviewed, 12 unresolved discussions (waiting on @ssheorey and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 222 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJKF7BSdedMR0piGQ:-MPmNHi7C1BlkwzKKHo3:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L222)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Could you also add a new factory function for "Create a 0-D tensor"? 

In numpy, this is:
```python
np.array(3.14)
```
</blockquote></details>

Done.
___
*[cpp/open3d/core/Tensor.h, line 254 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJg4t5XArzoMrwlep:-MPmNJaoCT2oVgC-IY-f:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L254)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

please reformat it to:`sequences (nested lists ...` (change space and line break location)
</blockquote></details>

Done.
___
*[cpp/tests/core/Tensor.cpp, line 135 at r6](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPh2aq11nI-31cTPNqh:-MPi8XSK2h5BxfdaBP9C:b-129m5o) ([raw file](https://github.com/intel-isl/open3d/blob/25251b1aead2e079297e1fc86a53ca9bd3b9f924/cpp/tests/core/Tensor.cpp#L135)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The initializer_list is unable to figure out dimension for 0-sized input and was giving error. I've added code to handle it in the 3D code.
___
*[cpp/tests/core/Tensor.cpp, line 144 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJya6AwIuL-bj26rB:-MPmMpvl-Ww6Su6eDp1X:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/tests/core/Tensor.cpp#L144)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Assert the shape is `{0,}`.
</blockquote></details>

Done.
___
*[cpp/tests/core/Tensor.cpp, line 148 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkKCwh6mVMz5lbltNQ:-MPmMsWIAqNnsJEtQpCG:bz9ply9) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/tests/core/Tensor.cpp#L148)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Assert the shape is `{2, 0}`.

Also, add shape and dtype assertion for other tests. The dtype assertion is needed to test if the template is converted correctly to core::Dtype.
</blockquote></details>

I thought, this was covered when we convert the tensor to flat vector so didn't add. I've added explicit tests for the same

**Review by yxlao — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: 1 of 2 files reviewed, 17 unresolved discussions (waiting on @ShubhamAgarwal12, @ssheorey, and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 254 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJg4t5XArzoMrwlep:-MPnHJmOBz4sBaQTKT-n:b-i27fd9) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L254)):*
<details><summary><i>Previously, ShubhamAgarwal12 (Shubham Agarwal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

we can change the line break to make it look prettier, it will look like this and still fit in 80-chars limit:



same for the others.
___
*[cpp/tests/core/Tensor.cpp, line 106 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnHlpt2-nKOcF7tkCM:-MPnHlpt2-nKOcF7tkCN:b-qzfxiq) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L106)):*
> ```cpp
>     core::Device device = GetParam();
> 
>     // 0-D tesnor with given value
> ```

Add `.`
___
*[cpp/tests/core/Tensor.cpp, line 107 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnHqPN7tfzMOypk5E4:-MPnHqPN7tfzMOypk5E5:b4roghf) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L107)):*
> ```cpp
> 
>     // 0-D tesnor with given value
>     std::vector<float> vals0{1};
> ```

Let's try to make the code cleaner. Merge this line to:
```cpp
EXPECT_EQ(t0.ToFlatVector<float>(), std::vector<float>({`}));
```
___
*[cpp/tests/core/Tensor.cpp, line 108 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnI4nc9nEtBl5WhAe_:-MPnI4nc9nEtBl5WhAea:b-6wjnhs) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L108)):*
> ```cpp
>     // 0-D tesnor with given value
>     std::vector<float> vals0{1};
>     core::Tensor t0 = core::Tensor::Init<float>(1, device);
> ```

Put `core::Tensor t;` at the very beggining. Then, simply do
```cpp
t = core::Tensor::Init<float>(1, device);
```

So that we don't need `t0, t1, t2, ...`.
___
*[cpp/tests/core/Tensor.cpp, line 109 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnIKmv2W-0O9NSst0W:-MPnIKmv2W-0O9NSst0X:b-x3ctns) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L109)):*
> ```cpp
>     std::vector<float> vals0{1};
>     core::Tensor t0 = core::Tensor::Init<float>(1, device);
>     core::SizeVector shape0;
> ```

Same, merge this line to the `EXPECT_EQ` line to make the code cleaner.
___
*[cpp/tests/core/Tensor.cpp, line 170 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnI_5JEPn6uHHW7lg7:-MPnI_5K16gPBWJ-fKTP:b-uul7ft) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L170)):*
> ```cpp
>     core::Tensor t6 = core::Tensor::Init<double>({}, device);
>     core::SizeVector shape6{
>             0,
> ```

do it without `,`

**Review by ShubhamAgarwal12 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2796)* status: 0 of 2 files reviewed, 17 unresolved discussions (waiting on @ssheorey and @yxlao)

___
*[cpp/open3d/core/Tensor.h, line 254 at r7](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPkJg4t5XArzoMrwlep:-MPnxwY7D_53ppyntGdx:b-1o8w4g) ([raw file](https://github.com/intel-isl/open3d/blob/e1ae52dddbaa387afe972b8b0ffce342af33f96c/cpp/open3d/core/Tensor.h#L254)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

we can change the line break to make it look prettier, it will look like this and still fit in 80-chars limit:



same for the others.
</blockquote></details>

I did it for the 1st dimension. For the 3D inner dimension, it exceeds the line limit, but now it seems better.
___
*[cpp/tests/core/Tensor.cpp, line 106 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnHlpt2-nKOcF7tkCM:-MPo0P9YC5araQRsozup:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L106)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Add `.`
</blockquote></details>

Done.
___
*[cpp/tests/core/Tensor.cpp, line 107 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnHqPN7tfzMOypk5E4:-MPnyMgt3dpJNqacwFxt:b-es3gw) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L107)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Let's try to make the code cleaner. Merge this line to:
```cpp
EXPECT_EQ(t0.ToFlatVector<float>(), std::vector<float>({`}));
```
</blockquote></details>

I've added the vector inline for the checks, wherever it was within the line limits. For others, I still use a separate variable.
___
*[cpp/tests/core/Tensor.cpp, line 108 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnI4nc9nEtBl5WhAe_:-MPnybTxCupeMgwZEquC:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L108)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Put `core::Tensor t;` at the very beggining. Then, simply do
```cpp
t = core::Tensor::Init<float>(1, device);
```

So that we don't need `t0, t1, t2, ...`.
</blockquote></details>

Done.
___
*[cpp/tests/core/Tensor.cpp, line 109 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnIKmv2W-0O9NSst0W:-MPo0Tyl7luN-4-Z3zGH:b-hu8y1q) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L109)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

Same, merge this line to the `EXPECT_EQ` line to make the code cleaner.
</blockquote></details>

Merged for all the shapes.
___
*[cpp/tests/core/Tensor.cpp, line 170 at r8](https://reviewable.io/reviews/intel-isl/open3d/2796#-MPnI_5JEPn6uHHW7lg7:-MPo0YBR0ldB8i0XLaT6:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/f8c2018449f2b7b557927499504fca7371a36fc1/cpp/tests/core/Tensor.cpp#L170)):*
<details><summary><i>Previously, yxlao (Yixing Lao) wrote…</i></summary><blockquote>

do it without `,`
</blockquote></details>

Done.

**Review by yxlao — APPROVED** [review]

**Pull request merged**